datafusion/physical-plan/src/stream.rs

```diff
@@ -472,6 +498,7 @@ impl Stream for ObservedStream {
         cx: &mut Context<'_>,
     ) -> Poll<Option<Self::Item>> {
         let poll = self.inner.poll_next_unpin(cx);
+        let poll = self.limit_reached(poll);
```

Comment:
Nit:

> let Some(fetch) = self.fetch else { return poll };

May be we can move the following to above, only call  limit_reached when  
```rust
let Some(fetch) = self.fetch {
    poll = self.limit_reached(poll);
}
```